src/tools/remote-test-client/src/main.rs

```diff
@@ -116,11 +165,14 @@ fn push(path: &Path) {
     let client = t!(TcpStream::connect("127.0.0.1:12345"));
     let mut client = BufWriter::new(client);
     t!(client.write_all(b"push"));
-    t!(client.write_all(path.file_name().unwrap().to_str().unwrap().as_bytes()));
-    t!(client.write_all(&[0]));
-    let mut file = t!(File::open(path));
-    t!(io::copy(&mut file, &mut client));
+    send(path, &mut client);
     t!(client.flush());
+
+    // Wait for an acknowledgement that all the data was received. No idea
+    // why this is necessary, seems like it shouldn't be!
```

Comment:
Just to clarify: without the server sending and us waiting on the `"ack "` the data is not properly received?

My best guess is that the immediate `close` of the socket after sending the data could cause problems, which seems to indeed be the case see: http://blog.csdn.net/CPP_CHEN/article/details/29864509